leptonai/util/cancel_on_disconnect.py

```diff
@@ -0,0 +1,101 @@
+"""
+Implements the functionality to cancel a fastapi call when the client disconnects.
+"""
+
+import anyio
+from exceptiongroup import catch, ExceptionGroup
+from fastapi import Request, HTTPException
+from loguru import logger
+
+
+class ClientDisconnected(Exception):
+    """
+    Exception raised when the client disconnects.
+    """
+
+    pass
+
+
+class TaskFinished(Exception):
+    """
+    Exception raised when the task finishes.
+    """
+
+    pass
+
+
+async def _disconnect_poller(request: Request, interval: float):
+    """
+    Poll for disconnect. If the request disconnects, stop polling and return.
+    """
+    logger.trace("Starting disconnect poller")
+    try:
+        while True:
+            await anyio.sleep(interval)
+            if await request.is_disconnected():
+                logger.trace(f"Request {request} disconnected.")
+                raise ClientDisconnected()
+    except anyio.get_cancelled_exc_class():
+        logger.trace("Stopping disconnect poller")
+
+
+async def run_with_cancel_on_disconnect(
+    request: Request, cancel_on_connect_interval: float, callback, *args, **kwargs
+):
+    """
+    Run the callback with the given arguments. If the client disconnects, cancel the callback.
+
+    Args:
+        request: The request object.
+        cancel_on_connect_interval: The interval in seconds to poll for disconnect.
+        callback: The callback to run.
+        args: The arguments to pass to the callback.
+        kwargs: The keyword arguments to pass to the callback.
+    """
+    result = []
+
+    async def _callback_task_wrapper():
+        logger.trace("Starting callback task wrapper")
+        try:
+            ret = await callback(*args, **kwargs)
+            result.append(ret)
+            raise TaskFinished()
+        except anyio.get_cancelled_exc_class():
```

Comment:
if the callback gets cancelled before line 60 finishes, result will be empty list, and it will cause error at line 100 and 101?

Reply:
Basically, if the callback gets cancelled before line 60 (actually, anywhere before line 62), then it means that we are going to handle the ClientDisconnected exception, and line 91's `with` statement will trigger `handle_client_disconnected`, which throws the `HTTPException`, meaning that line 100 and 101 won't be executed, so it won't be an error.

In fact, line 100 and 101 will only be triggered if the task finished successfully and did not raise any user error. There are three cases:
- If the task finishes successfully, `handle_task_finished` will be a no-op (other than programmatically check if result has length 1), and the execution moves on to line 100 and 101.
- If the client disconnecte, `handle_client_disconnected` will raise a HTTPException up.
- If the user code has other exceptions, the `_callback_task_wrapper` will convert it as a HTTPException (line 70-73), which will then trigger the `handle_http_exception` function, which further raises it up.